core/src/test/java/org/elasticsearch/discovery/zen/NodeJoinControllerTests.java

```diff
@@ -413,6 +413,15 @@ public void testNewClusterStateOnExistingNodeJoin() throws InterruptedException,
         assertTrue("failed to publish a new state upon existing join", clusterService.state() != state);
     }
 
+    public void testRejectingJoinWithSameAddress() throws InterruptedException, ExecutionException {
+        ClusterState state = clusterService.state();
+        final DiscoveryNode other_node = new DiscoveryNode("other_node", state.nodes().getLocalNode().getAddress(),
+            emptyMap(), emptySet(), Version.CURRENT);
+
+        ExecutionException e = expectThrows(ExecutionException.class, () -> joinNode(other_node));
+        assertThat(e.getMessage(), containsString("found existing node"));
+    }
+
```

Comment:
also add test for situation I described above (same node id but different address)